packages/widgets/src/blocks/legacy-widget/edit/utils.js

```diff
@@ -0,0 +1,104 @@
+/**
+ * WordPress dependencies
+ */
+import { useRef, useLayoutEffect, useState } from '@wordpress/element';
+import { useResizeObserver } from '@wordpress/compose';
+
+/**
+ * Positions an element by the top of an objective element if it can fit or
+ * otherwise as close as possible within the window.
+ *
+ * @param {Element} objective    Element to take position from.
+ * @param {Object}  $1           Options.
+ * @param {boolean} $1.isEnabled Whether to apply the hook or not.
+ *
+ * @return {Object} A ref and resize observer element.
+ */
+export function useBudgeTopBy( objective, { isEnabled } ) {
```

Comment:
It's a minor thing, but I feel like the naming could be iterated on. If this is just for use on the widget form it could take naming from that. `useWideWidgetPopoverPositioning` or if this could become something more generic `useVerticalPopoverPositioning`.

Reply:
I tried another name intended to describe the action `useAlignTopWithinViewport` but I'd be fine with changing to something like your first suggestion as this isn't intended to for other applications. I do like the idea that this hook could be removed at some point in favor of something more generic.